pkg/cloudprovider/provider/baremetal/plugins/tinkerbell/client/template.go

```diff
@@ -20,73 +20,252 @@ import (
-		if err.Error() == sqlErrorString || err.Error() == sqlErrorStringAlt {
-			return nil, fmt.Errorf("template %w", ErrNotFound)
-		}
-
-		return nil, fmt.Errorf("getting template from Tinkerbell: %w", err)
+func (t *TemplateClient) Delete(ctx context.Context, namespacedName types.NamespacedName) error {
+	template := &tinkv1alpha1.Template{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      namespacedName.Name,
+			Namespace: namespacedName.Namespace,
+		},
+	}
+	// Create the Template object in the Tinkerbell cluster
+	if err := t.tinkclient.Delete(ctx, template); err != nil {
+		return fmt.Errorf("failed to delete Template in Tinkerbell cluster: %w", err)
 	}
 
-	return tinkTemplate, nil
+	return nil
 }
 
-// Update a Tinkerbell Template.
-func (t *Template) Update(ctx context.Context, template *template.WorkflowTemplate) error {
-	if _, err := t.client.UpdateTemplate(ctx, template); err != nil {
-		return fmt.Errorf("updating template in Tinkerbefll: %w", err)
+// CreateTemplate creates a Tinkerbell Template in the Kubernetes cluster.
+func (t *TemplateClient) CreateTemplate(ctx context.Context, hardware *tinkv1alpha1.Hardware, namespace, osImageURL string) error {
+	template := &tinkv1alpha1.Template{}
+	if err := t.tinkclient.Get(ctx, types.NamespacedName{
+		Name:      ProvisionWorkerNodeTemplate,
+		Namespace: namespace,
+	}, template); err != nil {
+		if kerrors.IsNotFound(err) {
+			data, err := getTemplate(hardware, osImageURL)
+			if err != nil {
+				return err
+			}
+
+			template.Name = ProvisionWorkerNodeTemplate
+			template.Namespace = namespace
+			template.Spec = tinkv1alpha1.TemplateSpec{
+				Data: &data, // templateData is a string containing the YAML definition.
+			}
+
+			// Create the Template object in the Tinkerbell cluster
+			if err := t.tinkclient.Create(ctx, template); err != nil {
+				return fmt.Errorf("failed to create Template in Tinkerbell cluster: %w", err)
+			}
+
+			return nil
+		}
+
+		return fmt.Errorf("failed to get template %s: %v", ProvisionWorkerNodeTemplate, err)
 	}
 
 	return nil
 }
 
-// Create a Tinkerbell Template.
-func (t *Template) Create(ctx context.Context, template *template.WorkflowTemplate) error {
-	resp, err := t.client.CreateTemplate(ctx, template)
+func getTemplate(hardware *tinkv1alpha1.Hardware, osImageURL string) (string, error) {
+	actions := []Action{
+		createWipeDiskAction(),
+		createStreamUbuntuImageAction(hardwareDisk1, osImageURL),
+		createGrowPartitionAction(hardwareDisk1),
+		createNetworkConfigAction(hardware),
+		createCloudInitConfigAction(),
+		decodeCloudInitFile(hardware.Name),
+	}
+
+	task := Task{
+		Name:       "os-installation",
+		WorkerAddr: "{{.device_1}}",
+		Volumes:    []string{"/dev:/dev", "/dev/console:/dev/console", "/lib/firmware:/lib/firmware:ro"},
+		Actions:    actions,
+	}
+
+	template := Template{
+		Name:          "ubuntu",
+		Version:       "0.1",
+		GlobalTimeout: 1800,
+		Tasks:         []Task{task},
+	}
+	// Marshal task to YAML
```

Comment:
This comment isn't adding anything and can be removed instead of fixed.

Reply:
This is part of the leftovers. Will be removed.